Spigot-Server-Patches/0418-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,525 @@
+                 if (event.isCancelled()) return true;
+@@ -1251,8 +1256,11 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             }
+         }
+ 
+-        for (int k = i - this.viewDistance; k <= i + this.viewDistance; ++k) {
+-            for (int l = j - this.viewDistance; l <= j + this.viewDistance; ++l) {
++        //Paper start ticket level pp-view-dist
++        int playerViewDistance = entityplayer.getTrackViewDistance();
++        for (int k = i - playerViewDistance; k <= i + playerViewDistance; ++k) {
++            for (int l = j - playerViewDistance; l <= j + playerViewDistance; ++l) {
++                //Paper end
+                 ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(k, l);
+ 
+                 this.sendChunk(entityplayer, chunkcoordintpair, new Packet[2], !flag, flag);
+@@ -1320,17 +1328,19 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+         int k1;
+         int l1;
+ 
+-        if (Math.abs(i1 - i) <= this.viewDistance * 2 && Math.abs(j1 - j) <= this.viewDistance * 2) {
+-            k1 = Math.min(i, i1) - this.viewDistance;
+-            l1 = Math.min(j, j1) - this.viewDistance;
+-            int i2 = Math.max(i, i1) + this.viewDistance;
+-            int j2 = Math.max(j, j1) + this.viewDistance;
++        //Paper start ticket level pp-view-dist
++        int playerViewDistance = entityplayer.getTrackViewDistance();
++        if (Math.abs(i1 - i) <= playerViewDistance * 2 && Math.abs(j1 - j) <= playerViewDistance * 2) {
++            k1 = Math.min(i, i1) - playerViewDistance;
++            l1 = Math.min(j, j1) - playerViewDistance;
++            int i2 = Math.max(i, i1) + playerViewDistance;
++            int j2 = Math.max(j, j1) + playerViewDistance;
+ 
+             for (int k2 = k1; k2 <= i2; ++k2) {
+                 for (int l2 = l1; l2 <= j2; ++l2) {
+                     ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(k2, l2);
+-                    boolean flag3 = a(chunkcoordintpair, i1, j1) <= this.viewDistance;
+-                    boolean flag4 = a(chunkcoordintpair, i, j) <= this.viewDistance;
++                    boolean flag3 = a(chunkcoordintpair, i1, j1) <= playerViewDistance;
++                    boolean flag4 = a(chunkcoordintpair, i, j) <= playerViewDistance;
+ 
+                     this.sendChunk(entityplayer, chunkcoordintpair, new Packet[2], flag3, flag4);
+                 }
+@@ -1340,8 +1350,8 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+             boolean flag5;
+             boolean flag6;
+ 
+-            for (k1 = i1 - this.viewDistance; k1 <= i1 + this.viewDistance; ++k1) {
+-                for (l1 = j1 - this.viewDistance; l1 <= j1 + this.viewDistance; ++l1) {
++            for (k1 = i1 - playerViewDistance; k1 <= i1 + playerViewDistance; ++k1) {
++                for (l1 = j1 - playerViewDistance; l1 <= j1 + playerViewDistance; ++l1) {
```

Comment:
missing paper comment

Reply:
the //paper start encompasses this whole region to line 422 in the patch at //paper end

Reply:
although it may be better to have it in small sections

Reply:
I know, it's too large of a block to comment
The comments are there to tell us what lines we have changed, it's very useful when solving diffs

Reply:
yes its been changed